### ISSUE TITLE
Issues with Handling File Inputs and Image Previews in Next.js 15 Form Actions

### ISSUE THREAD

**Issue by Marabii** [vercel/next.js#72958]
### Link to the code that reproduces this issue

https://github.com/Marabii/next-js-issue

### To Reproduce

**Minimal Reproducible Example:**

*Client Component (`Home.tsx`):*
```tsx
"use client";

import { useActionState } from "react";
import { useState } from "react";
import sendDataAction, { State } from "./action";
import Image from "next/image";

export default function Home() {
  const [state, formAction, pending] = useActionState<State, FormData>(
    sendDataAction,
    null
  );
  const [selectedImages, setSelectedImages] = useState<File[]>([]);

  const handleImageChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files) {
      setSelectedImages(Array.from(event.target.files));
    }
  };

  const handleImageRemove = (index: number) => {
    setSelectedImages((prevImages) =>
      prevImages.filter((_, imgIndex) => imgIndex !== index)
    );
  };

  return (
    <div className="flex items-center justify-center min-h-screen bg-gray-100">
      <form
        action={formAction}
        className="bg-white shadow-md rounded-lg p-8 space-y-6 w-full max-w-md"
      >
        <div>
          <label
            htmlFor="images"
            className="block text-sm font-medium text-gray-700 mb-2"
          >
            Upload Images
          </label>
          <input
            type="file"
            name="images"
            id="images"
            multiple
            onChange={handleImageChange}
            required
            className="block w-full text-sm text-gray-900 border border-gray-300 rounded-lg cursor-pointer bg-gray-50 focus:outline-none focus:ring focus:ring-indigo-500 focus:border-indigo-500"
          />
          <div className="mt-4">
            <h2 className="text-sm font-medium text-gray-800 mb-2">
              Selected Images
            </h2>
            <div className="grid grid-cols-2 gap-2">
              {selectedImages.map((image, index) => (
                <div
                  key={index}
                  className="relative w-full h-24 bg-gray-100 rounded overflow-hidden border border-gray-200"
                >
                  <Image
                    src={URL.createObjectURL(image)}
                    alt={`Selected image ${index + 1}`}
                    layout="fill"
                    objectFit="cover"
                    className="rounded"
                  />
                  <button
                    type="button"
                    onClick={() => handleImageRemove(index)}
                    className="absolute top-1 right-1 bg-red-500 text-white rounded-full p-1 hover:bg-red-600 focus:outline-none"
                  >
                    ✕
                  </button>
                </div>
              ))}
            </div>
          </div>
        </div>
        <button
          disabled={pending}
          type="submit"
          className={`w-full py-2 px-4 rounded-lg text-white ${
            pending
              ? "bg-gray-400 cursor-not-allowed"
              : "bg-indigo-600 hover:bg-indigo-700 focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
          }`}
        >
          {pending ? "Submitting..." : "Submit"}
        </button>
        <div className="text-sm text-gray-500 mt-4">
          <pre>{JSON.stringify(state, null, 2)}</pre>
        </div>
      </form>
    </div>
  );
}
```

*Server Action (`action.ts`):*
```ts
"use server";

const sendDataAction = async (
  state: State,
  formData: FormData
): Promise<State> => {
  // Simulating a validation error
  return {
    status: "failure",
    errors:
      "This isn't a real error, but just simulating an error that could occur in a big form with multiple inputs",
  };
};

export default sendDataAction;

export type State = {
  status: "success" | "failure";
  errors?: string;
} | null;
```

**Steps to Reproduce:**

1. Create a Next.js 15 application with the above client and server components.
2. Run the application and navigate to the form.
3. Select multiple image files using the file input.
4. Click the "Submit" button to trigger the form submission.
5. Observe that:
   - The file input resets, requiring re-selection of images.
   - The image previews remain displayed due to the `selectedImages` state.
   - If attempting to remove an image from the preview, the file input does not reflect this change.

### Current vs. Expected behavior

**Description:**  
I'm experiencing challenges when working with file inputs, specifically image uploads, in a Next.js 15 application using form actions. The issues arise when attempting to preview selected images and implement a remove functionality. Below are the detailed problems and a minimal reproducible example.

**Issues Encountered:**

1. **Image Inputs Reset on Submit with Validation Errors:**
   - When the form is submitted, the file input resets, causing the user to reselect images if there are validation errors handled in the server action.
   - Despite the file input resetting, the `selectedImages` state on the client side still holds the previously selected images and displays the previews.
   - This discrepancy leads to confusion as the user sees the previews but is required to re-upload the images.

2. **Inability to Synchronize File Input with Selected Images State:**
   - It's not possible to programmatically set the value of the file input (e.g., `value={selectedImages}`) due to security restrictions.
   - This limitation prevents synchronizing the file input with the `selectedImages` state, especially when users remove images from the preview. The file input remains unchanged, leading to inconsistencies.

**Impact:**  
These issues hinder the user experience by creating confusion between the displayed image previews and the actual state of the file input. Users may find it frustrating to re-upload images after encountering validation errors, and the inability to manage the file input state programmatically limits the functionality of the form.

**Expected Behavior:**

- **On Validation Errors:**
  - The file input should retain the selected files to prevent users from re-uploading them.
  - The image previews should accurately reflect the current state of the file input.

- **Synchronizing File Input with State:**
  - It should be possible to programmatically manage the file input's value to match the `selectedImages` state. I know it's not currently possible due to security reasons, but there has to be a workaround.

### Provide environment information

```bash
Operating System:
 Platform: linux mint 22 cinamon
Relevant packages:
 next: 15.0.3
 react and react-dom: 19.0.0-rc-66855b96-20241106
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), next start (local), Vercel (Deployed), Other (Deployed)

**Comment by master4o** [comment]
model of server actions is broken, its reset form everytime even when form action return errors (in docs Next recommend to return errors in objects - https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#server-side-form-validation). 

As for me, i decide to use formData cycle for error state. In server action next passes formData we can return object with formData from server action and can pass values in inputs after. Next resets value, we returns them back :)

**Comment by Marabii** [comment]
Thank you for your response! Your idea is excellent—I can’t believe I didn’t think of it myself. Unfortunately, it doesn’t resolve the issue I’m facing. While I do have access to the formData object after validation, I’m unable to programmatically push file data into the input due to security restrictions. This limitation is particularly challenging for file inputs. For other input types, I’ve worked around it by creating a custom input component that stores the input data in state and programmatically updates it.

**Comment by master4o** [comment]
@Marabii, try to add state.formData.get('images') in defaultValue prop

**Comment by Marabii** [comment]
Thanks for trying to help me but it doesn't work. I really think there is no way to do this unless they developed a specific Input file component for ths usecase.

**Comment by liri2006** [comment]
@Marabii in order to preserve file you can use useState hook and then manually append it to formData on form submit:

```
export default function AccountForm() {
  const [selectedFile, setSelectedFile] = useState<File | null>(null)
  const [formState, formAction, isPending] = useActionState(saveMyAccount, null)

  const handleSubmit = (formData: FormData) => {
      if (selectedFile) formData.set('apiKey', selectedFile)
      formAction(formData)
  }

  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files?.[0] || null
    setSelectedFile(file)
  }

  return (
    <form action={handleSubmit} className="mx-auto max-w-4xl">
       <input
          id="dropzone-file"
          name="apiKey"
          type="file"
          onChange={handleFileChange}
        />
     </form>
   )
}
```

**Comment by Marabii** [comment]
Your idea works, thanks a lot but I hope there was a better way of doing it. I didn't want to mess with the formData object but i guess there is no other solution. I'm closing this issue, thanks a lot @liri2006 and @master4o

**Comment by liri2006** [comment]
@Marabii yeah, I was also disappointed, it seems like useState-less form handling is what they are after here with serverActions, but it supports only basic use cases with plain inputs for now. Even select does not work without additional hassle.  
Maybe keep issue open, so developers know about these shortcomings?

**Comment by Marabii** [comment]
I reopened this issue per @liri2006 request, it seems like manual manipulation of the formdata object is mandatory in cases like file inputs and select inputs.

**Comment by LeonN534** [comment]
As the [Next js docs](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#passing-additional-arguments) say, you do not need to modify the formData object, you can use the bind() method of the server action. 

**The form (client component)**

```
  const [photo, setPhoto] = useState<File | null>(null);
  const [state, formAction, isLoading] = useActionState(
    createTeacherAction.bind(null, photo),
    initialState,
  );

  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files?.[0];
    if (file) {
      setPhoto(file);
      setPhotoUrl(URL.createObjectURL(file));
    }
  };


<input
  type="file"
  accept="image/*"
  className="hidden"
  id="profilePhoto"
  name="profilePhoto"
  onChange={handleFileChange}
/>
```

**The server action**
```
export async function createTeacherAction(
  photo: File | null,
  prevState: FormState<CreateTeacherFields>,
  formData: FormData,
) {
  console.log("selectedFile", photo);
...
}
```

The only caveat I could find is that you will get an additional field on your form data that you won't use, since it is the file being sent in the file input.

**Comment by Marabii** [comment]
You're right, I didn't know the .bind JavaScript method, thanks for sharing your solution. In my project I created a Form wrapper component

**Comment by Marabii** [comment]
You're absolutely right—I wasn't familiar with the .bind method in JavaScript. Thank you for sharing your solution! To address similar needs in my project, I developed a FormWrapper component that extends the functionality of the standard <form> tag by incorporating server actions. Below is an overview of how it works, along with the relevant code snippets.

FormWrapper Component

The FormWrapper component serves as a context provider, allowing child components to access and manipulate additional form data seamlessly. It handles form submission, data merging, client-side validation, and redirects upon successful submission.

"use client";

import React, {
  useEffect,
  useState,
  createContext,
  useActionState,
} from "react";
import { useRouter } from "next/navigation";
import {
  ActionReturn,
  AdditionalDataContextType,
  FormWrapperProps,
} from "@/interfaces/Form/Form";
import { Errors } from "@/interfaces/Errors/errors";

// Initialize the context with the combined ActionReturn
export const AdditionalDataContext = createContext<
  AdditionalDataContextType<any>
>({
  additionalData: new FormData(),
  addAdditionalData: (formData: FormData) => {},
  actionReturn: {} as ActionReturn<any>,
  pending: false,
});

export default function FormWrapper<T>({
  children,
  action,
  redirectTo,
  className,
}: FormWrapperProps<T>) {
  const router = useRouter();
  const [additionalData, setAdditionalData] = useState<FormData>(
    new FormData()
  );

  // Local state for client-side errors
  const [localActionReturn, setLocalActionReturn] = useState<
    ActionReturn<T> | undefined
  >(undefined);

  const [actionReturn, formAction, pending] = useActionState<
    ActionReturn<T>,
    FormData
  >(action, null);

  // Redirect user to another directory upon successful server action
  useEffect(() => {
    if (!pending && actionReturn?.status === "success" && redirectTo) {
      console.log("rerouting the user...");
      router.replace(redirectTo);
    }
  }, [pending, actionReturn, redirectTo, router]);

  // Combine server and client action returns, prioritizing client errors
  const combinedActionReturn: ActionReturn<T> = mergeActionReturns(
    actionReturn,
    localActionReturn
  );

  const handleSubmit = async (formData: FormData) => {
    // Reset local errors before validation
    setLocalActionReturn({
      status: "success",
      data: {} as T,
      errors: undefined,
    });

    const mergedData: FormData = mergeFormData(formData, additionalData);
    const sizeInMB = getFormDataSizeInMB(mergedData);

    if (sizeInMB >= 2) {
      // Set client-side error without sending data to the backend
      setLocalActionReturn({
        status: "failure",
        data: {} as T,
        errors: {
          global: "Images are too large. Maximum allowed size is 2MB.",
        } as Errors<T>, // Explicitly cast to Errors<T>
      });
      return;
    }

    // Proceed with the form action if size is acceptable
    formAction(mergedData);
  };

  const addAdditionalData = (formData: FormData) => {
    setAdditionalData(mergeFormData(formData, additionalData));
  };

  return (
    <AdditionalDataContext.Provider
      value={{
        additionalData,
        addAdditionalData,
        actionReturn: combinedActionReturn,
        pending,
      }}
    >
      <form
        action={handleSubmit}
        className={
          className || "max-w-2xl mx-auto p-6 bg-white rounded-lg shadow-md"
        }
      >
        {children}
      </form>
    </AdditionalDataContext.Provider>
  );
}

function mergeFormData(formData1: FormData, formData2: FormData) {
  const mergedFormData = new FormData();

  // Append all entries from formData1
  for (const [key, value] of formData1.entries()) {
    mergedFormData.append(key, value);
  }

  // Append all entries from formData2
  for (const [key, value] of formData2.entries()) {
    mergedFormData.append(key, value);
  }

  return mergedFormData;
}

function mergeActionReturns<T>(
  actionReturn: ActionReturn<T>,
  localActionReturn: ActionReturn<T> | undefined
): ActionReturn<T> {
  if (!localActionReturn) {
    return actionReturn;
  }

  if (!actionReturn && localActionReturn) {
    return localActionReturn;
  }

  if (!actionReturn) {
    return null;
  }

  // If localActionReturn has errors undefined, it means errors have been reset
  const errors = {
    ...(actionReturn.errors || {}),
    ...(localActionReturn.errors || {}),
  };

  const status =
    actionReturn.status === "failure" || localActionReturn.status === "failure"
      ? "failure"
      : "success";

  const returnObj: ActionReturn<T> = {
    status: status,
    errors: errors as Errors<T>,
    data: { ...actionReturn.data, ...localActionReturn.data },
  };

  return returnObj;
}

function getFormDataSizeInMB(formData: FormData) {
  let totalSize = 0;

  // Iterate through each entry in the FormData
  for (let entry of formData.entries()) {
    const [_, value] = entry;

    if (value instanceof File) {
      // If the value is a File, use its size property
      totalSize += value.size;
    } else {
      // If it's a string, calculate its size in bytes
      totalSize += new Blob([value]).size;
    }
  }

  // Convert bytes to megabytes and return the result
  return totalSize / (1024 * 1024);
}

FormImagesInput Component

This component consumes the AdditionalDataContext to handle image uploads. It manages selected images, provides previews, and ensures that the combined form data doesn't exceed the specified size limit.

"use client";

import { useState, useEffect } from "react";
import useAdditionalData from "../../../../components/Form/useAdditionalData";

export default function FormImagesInput() {
  const [selectedImages, setSelectedImages] = useState<File[]>([]);
  const { addAdditionalData } = useAdditionalData();

  const handleImageUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files) {
      const filesArray = Array.from(event.target.files);
      setSelectedImages((prevImages) => [...prevImages, ...filesArray]);
    }
  };

  const handleImageDelete = (index: number) => {
    setSelectedImages((prevImages) => prevImages.filter((_, i) => i !== index));
  };

  useEffect(() => {
    const formData = new FormData();
    selectedImages.forEach(
      (file) => file.size && formData.append("images", file)
    );
    addAdditionalData(formData);
  }, [selectedImages]);

  return (
    <>
      <label className="block mb-4">
        <span className="text-gray-600">
          Upload images of the product you're ordering
        </span>
        <input
          type="file"
          name="selectedFiles"
          accept="image/*"
          multiple
          onChange={handleImageUpload}
          className="mt-1 block w-full p-2 border rounded border-gray-300"
        />
      </label>

      <div className="mt-4 flex flex-wrap gap-4">
        {selectedImages.map((image, index) => (
          <div
            key={index}
            className="relative w-32 h-32 border border-gray-300 rounded overflow-hidden"
          >
            <img
              src={URL.createObjectURL(image)}
              alt={`Preview ${index + 1}`}
              className="w-full h-full object-cover"
            />
            <button
              onClick={() => handleImageDelete(index)}
              className="absolute top-1 right-1 bg-red-500 text-white rounded-full w-6 h-6 flex items-center justify-center"
              aria-label="Delete image"
            >
              &times;
            </button>
          </div>
        ))}
      </div>
    </>
  );
}

How It Works

1. Context Setup: The AdditionalDataContext provides a way to share additional form data and action states across components. This is particularly useful for handling complex forms with multiple data sources.


2. Form Submission Handling: The FormWrapper intercepts the form submission process. It merges the main form data with any additional data provided by child components, such as uploaded images. Before submitting, it performs a client-side validation to ensure that the total size of the form data doesn't exceed 2MB. If it does, it sets a local error state to inform the user.


3. Action Returns Merging: The component intelligently merges server-side and client-side action returns. This ensures that client-side validations take precedence over server responses, providing a better user experience.


4. Image Uploads: The FormImagesInput component allows users to upload multiple images. It manages the state of selected images, provides real-time previews, and enables users to remove any unwanted images. The selected images are then added to the form data through the context.


5. Redirect on Success: Upon a successful form submission, the user is automatically redirected to a specified URL, enhancing the flow of the application.


This setup provides a robust and flexible way to manage complex forms with additional data requirements and file upload.

I'm open to any constructive criticism or suggestions to further improve this implementation. Your feedback would be greatly appreciated!
